### PR TITLE
Xunit attempt

### PR DESCRIPTION
**Pull request by Pauliukas7**
Will leave more comments in slack

### CODE REVIEW

File: ChessBoard.UnitTests/ChessBoardTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
+using System.Text;
+using Microsoft.Extensions.Logging;
+using Moq;
+
+namespace ChessBoardApp.ChessBoard.Tests
+{
+    class TestLogger<T> : ILogger<T>
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardTests.cs]: No need to test logger, I'd suggest removing it if we're not using

File: ChessBoard.UnitTests/ChessBoardTests.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            Exception exception,
+            Func<TState, Exception, string> formatter
+        )
+        {
+            string logMessage = formatter(state, exception);
+            LoggedMessages.Add(logMessage);
+        }
+    }
+
+    public class ChessBoardGeneratorTests
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardTests.cs]: Let's move into separate file - ChessBoardGenerator.Tests

File: ChessBoard.UnitTests/ChessBoardTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        [Fact]
+        public void GenerateRectangleBoard_ShouldLogCorrectChessBoard()
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardTests.cs]: Good attempt, but let's consider https://adform.slack.com/archives/C05ECTFN5GE/p1688356759903009?thread_ts=1688335639.652929&cid=C05ECTFN5GE points 2 and 3

File: ChessBoard.UnitTests/ChessBoardTests.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            // Act
+            chessBoardGenerator.GenerateRectangeBoard(width, length);
+
+            // Assert
+            string expectedChessBoard = GetExpectedChessBoard(width, length);
+            Assert.Contains(expectedChessBoard, string.Join("\r\n", logger.LoggedMessages));
+        }
+
+        private static string GetExpectedChessBoard(int width, int length)
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardTests.cs]: Good use. I'd think about moving private methods underneath all the public ones. That's generally the case, with an exception when private method is used once and closely related to public method.

File: ChessBoard.UnitTests/ChessBoardTests.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                }
+
+                chessBoard.AppendLine();
+            }
+
+            return chessBoard.ToString();
+        }
+    }
+
+    public class ChessBoardDialogTests
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardTests.cs]: Let's not test dialog as it would become increasingly complicated when testing console without some interface wrapper. We could, on the other hand do some validator tests, for example https://adform.slack.com/archives/C05ECTFN5GE/p1688357426508479?thread_ts=1688335639.652929&cid=C05ECTFN5GE point 6. 
You could provide multiple inputs and assert that ValidationResult is { IsValid == false, ErrorMessage == "validation error message" } and if input is correct - ValidationResult is { IsValid == true, ErrorMessage == null }

Or maybe you have some other suggestions, let's discuss ;)

File: ChessBoard.UnitTests/ChessBoard.UnitTest.csproj [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.2" />
+    <PackageReference Include="Moq" Version="4.18.4" />
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoard.UnitTest.csproj]: If we're not using it the reference is not needed, maybe I'll plan some additional tests when we move to api :)
Note to self - topics to be touched - dependency injection, swagger, moq

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+﻿using Moq;
+using Serilog;
+
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs:4]: Please let's take care of those whitespaces through the pr, emojis won't help here :D

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    public class ChessBoardGeneratorTests
+    {
+
+        [Theory]
+        [InlineData("1\r\n",1,1)]
+        [InlineData("101\r\n010\r\n101\r\n010\r\n", 3,4)]
+        [InlineData("1010\r\n0101\r\n1010\r\n0101\r\n", 4,4)]
+        [InlineData("1010\r\n0101\r\n1010\r\n0101\r\n1010\r\n0101\r\n", 4,6)]
+        public void GenerateRectangleBoard_ShouldReturnCorrectChessBoard(string expected, int width, int length)
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs:23]: For greater coverage let's copy-paste this and create 2 more tests for GenerateChessBoardOneLoop, main idea to have `var loggerMock = new Mock<ILogger>();`  as private field within class `ChessBoardGeneratorTests` and reuse it in those 2 tests.

Question - can we potentially move something else to private field?

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+﻿using Moq;
+using Serilog;
+
+
+
+namespace ChessBoardApp.ChessBoard.Tests
```
Comment (zaLTys) [inline ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs]: Namespace does not seem correct, should be `ChessBoard.UnitTests`, same comment for other files

File: ChessBoard.UnitTests/InputValidationsTests.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    public class InputValidationTests
+    {
+
+        
+
+        [Theory]
+        [InlineData("5")]
+        [InlineData("gg")]
+        [InlineData("-5")]
+        public void ValidateInput_ShouldReturnCorrectObject(string input)
```
Comment (zaLTys) [inline ChessBoard.UnitTests/InputValidationsTests.cs]: Pretty good test! Suggestion - copy-paste and split cases into `valid` and `invalid` scenarios, basically 2 nearly identical tests where one would be `ValidInput_DoesNotReturn_ValidationError` and `InvalidInput_Returns_ValidationError` 

From maintenance perspective easier to decide what's being tested, also this allows to isolate tests from additional logic and conditionals `if (int.TryParse(input, out var x) && x > 0)`

Please share your opinion, we can always discuss any of these questions in slack ;)

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
+using Microsoft.Extensions.Logging;
+
+namespace ChessBoardApp.ChessBoard
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/ChessBoardDialog.cs:1]: Correct namespace, but please doublecheck ChessBoardApp project file which has incorrect root namespace leftover and adjust if needed `<RootNamespace>_nettask</RootNamespace>` => `<RootNamespace>ChessBoardApp</RootNamespace>`

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
+using Microsoft.Extensions.Logging;
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/ChessBoardDialog.cs]: Not used, tip - easy to remove these excessive usings by ctrl + K ctrl+E

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+            return Console.ReadLine();
+        }
+
+        public static string EnterRectangleLength()
+        {
+            Console.Write("Please enter rectangle length: ");
+            return Console.ReadLine();
+        }
+
+        public static void ChessBoardMessage()
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/ChessBoardDialog.cs:17]: Does not seem to be used, let's clean this up if that's the case

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                for (int j = 0; j < numberOfRows; j++)
+                {
+                    Console.Write((j + i) % 2 == 0 ? "1" : "0");
+                }
+                Console.WriteLine();
+            }
+        }
+
+        public string GenerateRectangeBoard(int width, int length)
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: typo in `Rectangle`

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+                {
+
+                    chessBoard.Append((j + i) % 2 == 0 ? "1" : "0");
+                }
+
+
+                chessBoard.AppendLine();
+            }
+
+            _logger.Information("Generated Board:" +Environment.NewLine + chessBoard.ToString());
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: formatting `+Environment.NewLine`  also quickfix with  `ctrl + K, ctrl + E `

File: ChessBoardApp/ChessBoard/InputValidation.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+    public class InputValidation
+    {
+        private readonly ILogger _logger;
+
+        public InputValidation(ILogger logger)
+        {
+            _logger = logger;
+        }
+        public class ValidationResult
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/InputValidation.cs:14]: Question - it is a fairly good use case to have `ValidationResult` as nested class inside `InputValidation` class. What was your reason for this approach? What could be other use cases?

File: ChessBoardApp/ChessBoard/InputValidation.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            public bool IsValid { get; set; }
+            public string ErrorMessage { get; set; }
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/InputValidation.cs]: We could use nullable string here to avoid warning messages

File: ChessBoardApp/ChessBoard/InputValidation.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+            bool isValid = int.TryParse(input, out var validInteger) && validInteger > 0;
+
+            if (!isValid)
+            {
+                string errorMessage = "Invalid input. An integer higher than 0 is required";
+                _logger.Error($"Invalid input '{input}'");
+                return new ValidationResult(false, errorMessage);
+            }
+
+            return new ValidationResult(true, null);
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoard/InputValidation.cs]: No need to adjust, just an insight - we can use static methods as constructors for validation result, then we  would not need to think what "true, null" means when reading the code. 
To do this you can change ctor for `ValidationResult` to private and add Success/Failure methods:
image

File: ChessBoardApp/ChessBoardApp.csproj [hunk 19]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <RootNamespace>_nettask</RootNamespace>
```
Comment (zaLTys) [inline ChessBoardApp/ChessBoardApp.csproj]: `<RootNamespace>ChessBoardApp</RootNamespace>`

**Review by zaLTys — CHANGES_REQUESTED** [review]

File: ChessBoard.UnitTests/ChessBoardTests.cs (hunk 5 above)
Comment (zaLTys) [reply ChessBoard.UnitTests/ChessBoardTests.cs]: well done!

File: ChessBoard.UnitTests/InputValidationsTests.cs [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+            if (int.TryParse(input, out var x) && x > 0)
+            {
+                Assert.True(validationResult.IsValid);
+                Assert.Null(validationResult.ErrorMessage);
+            }
+            else
+            {
+                Assert.False(validationResult.IsValid);
+                Assert.Equal("Invalid input. An integer higher than 0 is required", validationResult.ErrorMessage);
+            }
```
Comment (RaimondSob) [inline ChessBoard.UnitTests/InputValidationsTests.cs]: This is bad logic
In ValidateInput method you are generating validation result based on the same logic you are asserting:
`int.TryParse(input, out var x) && x > 0`
It would be the same as:
`if (validationResult.IsValid)
            {
                Assert.True(validationResult.IsValid);
                Assert.Null(validationResult.ErrorMessage);
            }
            `
If you split these test into positive and negative case you would avoid this

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+                {
+                    Console.Write((j + i) % 2 == 0 ? "1" : "0");
+                }
+                Console.WriteLine();
+            }
+        }
+
+        public string GenerateRectangeBoard(int width, int length)
+        {
+            StringBuilder chessBoard = new StringBuilder();
```
Comment (RaimondSob) [inline ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: You can use var here, not to write StringBuilder two times

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs (hunk 7 above)
Comment (Pauliukas7) [reply ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs:4]: Started using Moq in tests, so kept "using Moq"

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs (hunk 9 above)
Comment (Pauliukas7) [reply ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs]: adjusted

File: ChessBoard.UnitTests/InputValidationsTests.cs (hunk 10 above)
Comment (Pauliukas7) [reply ChessBoard.UnitTests/InputValidationsTests.cs]: Yes, this looks a lot cleaner:

```
public class InputValidationTests
    {
        private readonly Mock<ILogger> loggerMock;
        private readonly InputValidation inputValidation;

        public InputValidationTests()
        {
            loggerMock = new Mock<ILogger>();
            inputValidation = new InputValidation(loggerMock.Object);
        }

        [Theory]
        [InlineData("5")]
        public void ValidInput_DoesNotReturn_ValidationError(string input)
        {
            // Act
            var validationResult = inputValidation.ValidateInput(input);

            // Assert
            Assert.True(validationResult.IsValid);
            Assert.Null(validationResult.ErrorMessage);
        }

        [Theory]
        [InlineData("-5")]
        [InlineData("gg")]
        public void InvalidInput_Returns_ValidationError(string input)
        {
            // Act
            var validationResult = inputValidation.ValidateInput(input);

            // Assert
            Assert.False(validationResult.IsValid);
            Assert.Equal("Invalid input. An integer higher than 0 is required", validationResult.ErrorMessage);
        }
    }
```

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs (hunk 11 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardDialog.cs:1]: adjusted

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs (hunk 12 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardDialog.cs]: adjusted, shortcut remembered :D

File: ChessBoardApp/ChessBoard/ChessBoardDialog.cs (hunk 13 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardDialog.cs:17]: Should I leave one general message in the Dialog, for both Rectangle and Square chessboards? Because I had one for Rectangle and one for 4x4 5x5 etc. ChessBoards, so should I make one general message for all (`GenerateRectangleBoard`, `OneLoop`, `TwoLoops`) ?

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs (hunk 14 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: adjusted

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs (hunk 15 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: adjusted

File: ChessBoardApp/ChessBoard/InputValidation.cs (hunk 17 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/InputValidation.cs]: adjusted

File: ChessBoardApp/ChessBoard/InputValidation.cs (hunk 18 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/InputValidation.cs]: I liked the idea, so I adjusted it. As you've mentioned the ValidateInput method now only returns predefined objects - this suggests what kind of result is it (failure or success) and no need to think about what do the arguments (true,null) mean

File: ChessBoardApp/ChessBoardApp.csproj (hunk 19 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoardApp.csproj]: adjusted

File: ChessBoard.UnitTests/InputValidationsTests.cs (hunk 20 above)
Comment (Pauliukas7) [reply ChessBoard.UnitTests/InputValidationsTests.cs]: split the tests to positive and negative, adjusted

File: ChessBoardApp/ChessBoard/ChessBoardGenerator.cs (hunk 21 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/ChessBoardGenerator.cs]: adjusted

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs (hunk 8 above)
Comment (Pauliukas7) [reply ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs:23]: Maybe moving `chessBoardGenerator` instance? 
```
public class ChessBoardGeneratorTests
    {
        private readonly Mock<ILogger> loggerMock;
        private readonly ChessBoardGenerator chessBoardGenerator;

        public ChessBoardGeneratorTests()
        {
            loggerMock = new Mock<ILogger>();
            chessBoardGenerator = new ChessBoardGenerator(loggerMock.Object);
        }

        [Theory]
        [InlineData("1\r\n", 1, 1)]
        [InlineData("101\r\n010\r\n101\r\n010\r\n", 3, 4)]
        [InlineData("1010\r\n0101\r\n1010\r\n0101\r\n", 4, 4)]
        [InlineData("1010\r\n0101\r\n1010\r\n0101\r\n1010\r\n0101\r\n", 4, 6)]
        public void GenerateRectangleBoard_ShouldReturnCorrectChessBoard(string expected, int width, int length)
        {
            // Act
            var result = chessBoardGenerator.GenerateRectangleBoard(width, length);

            // Assert
            Assert.Contains(expected, result);
        }
```
I also created a test for both OneLoop and TwoLoops:

Let me know if I should get rid of any of them (talking about the method itself and the test) and also if moving the `chessBoardGenerator` instance to private field within class was correct option

File: ChessBoardApp/ChessBoard/InputValidation.cs (hunk 16 above)
Comment (Pauliukas7) [reply ChessBoardApp/ChessBoard/InputValidation.cs:14]: Maybe nesting is better because we are not planning to have any other classes than `InputValidation` so it would not make sense to have `ValidationResult` class as a separate public class, since all of the instances would only be created inside `InputValidation` ? Also having the `ValidationResult` as a separate public class, is more error-prone, whereas now you can only use it with `InputValidation` instances. While writing this I googled, and I see it's called **encapsulation**.

File: ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs (hunk 8 above)
Comment (zaLTys) [reply ChessBoard.UnitTests/ChessBoardGenerator.Tests.cs:23]: Good job,  now something to keep in mind - internal and private fields are prefixed by "_" by convention, let's uptade this with next step. Good for now.
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case
also, since we're not using loggerMock anywhere but in constructor, it could now be made into local variable within  that constructor:
image

Resolved.

File: ChessBoard.UnitTests/InputValidationsTests.cs (hunk 10 above)
Comment (zaLTys) [reply ChessBoard.UnitTests/InputValidationsTests.cs]: Could also test for empty string input, but looks good.

Resolved.

File: ChessBoardApp/ChessBoard/InputValidation.cs (hunk 16 above)
Comment (zaLTys) [reply ChessBoardApp/ChessBoard/InputValidation.cs:14]: Good answer, resolved.

**Review by zaLTys — APPROVED** [review]
